### PR TITLE
feat: add user favorite routes management and leaderboard functionality

### LINKED ISSUE #10: User Module - Favorites
Implement CRUD endpoints (Create, Read, Delete) for users to save and manage their favorite bus stops and routes.

### PR DESCRIPTION
**Pull request by LUCiv175**
- Implemented GET and POST endpoints for managing user's favorite routes.
- Added validation for favorite routes input, ensuring only integers are accepted and duplicates are removed.
- Created UserController methods to handle favorite routes retrieval and updates.
- Updated User model to include favoriteRoutes field.
- Added tests for favorite routes API and leaderboard functionality, ensuring proper authentication and response structure.
- Enhanced existing tests to cover edge cases for favorite routes and leaderboard ranking.

### CODE REVIEW

**Comment by LUCiv175** [comment]
> @LUCiv175 What about the other Copilot comments?
github posted later that comments and i didn't see them

File: backend/src/models/User.js [hunk 1]
```diff
@@ -16,6 +16,10 @@ const userSchema = new mongoose.Schema(
     picture: { type: String, required: false },
     refreshToken: { type: [String], default: [] },
     points: { type: Number, default: 0, index: true },
+    favoriteRoutes: {
+      type: [{ type: Number }],
```
[the remark this answers is not in the document]
Comment (LUCiv175) [reply backend/src/models/User.js]: done

File: backend/tests/users.favoriteRoutes.test.js [hunk 2]
```diff
@@ -0,0 +1,334 @@
...
+
+      const response = await request(app)
+        .post("/users/favorite-routes")
+        .set("Authorization", `Bearer ${authToken}`)
+        .send({ favoriteRoutes });
+
+      expect(response.status).toBe(200);
+      expect(response.body.favoriteRoutes).toContain(0);
+    });
+  });
```
[the remark this answers is not in the document]
Comment (LUCiv175) [reply backend/tests/users.favoriteRoutes.test.js]: done

File: backend/tests/users.favoriteRoutes.test.js [hunk 3]
```diff
@@ -0,0 +1,334 @@
...
+      const favoriteRoutes = [101, -1, 303];
+
+      const response = await request(app)
+        .post("/users/favorite-routes")
+        .set("Authorization", `Bearer ${authToken}`)
+        .send({ favoriteRoutes });
+
+      expect(response.status).toBe(200);
+      expect(response.body.favoriteRoutes).toContain(-1);
+    });
```
[the remark this answers is not in the document]
Comment (LUCiv175) [reply backend/tests/users.favoriteRoutes.test.js]: done

File: backend/src/controllers/UserController.js [hunk 4]
```diff
@@ -54,3 +54,73 @@ export const getLeaderboard = async (req, res, next) => {
...
+  } catch (error) {
+    if (error.name === "ValidationError") {
+      return res.status(400).json({
+        success: false,
+        message: error.message,
+      });
+    }
+    next(error);
+  }
+};
```
[the remark this answers is not in the document]
Comment (LUCiv175) [reply backend/src/controllers/UserController.js]: done

File: backend/tests/users.favoriteRoutes.test.js [hunk 5]
```diff
@@ -0,0 +1,334 @@
...
+  }
+
+  save = async () => {
+    usersById.set(this._id, this);
+    if (this.email) {
+      usersByEmail.set(String(this.email).toLowerCase(), this);
+    }
+    return this;
+  };
+}
```
[the remark this answers is not in the document]
Comment (LUCiv175) [reply backend/tests/users.favoriteRoutes.test.js]: done

**Review by Ceralex — APPROVED** [review]
Top commit and extended tests!
